### PR TITLE
[MRG] [python] setup.py custom install & sdist

### PR DESCRIPTION
**Pull request by wxchan**
#650

### CODE REVIEW

**Comment by wxchan** [comment]
@guolinke I finished `install` and `sdist` parts, seems the test fails at `bdist` part with a strange error.

File: python-package/setup.py [hunk 1]
```diff
@@ -3,94 +3,78 @@
...
 
     if build_sdist:
-        print("remove library when building source distribution")
-        if os.path.exists("./lightgbm/Release/"):
-            shutil.rmtree('./lightgbm/Release/')
-        if os.path.isfile('./lightgbm/lib_lightgbm.so'):
-            os.remove('./lightgbm/lib_lightgbm.so')
-    else:
+        open("./_IS_SOURCE_PACKAGE.txt", 'w').close()
+
```
Comment (wxchan) [inline python-package/setup.py]: L37-L58 is needed only if build_sdist is False?

Comment (wxchan) [inline python-package/setup.py]: and `--mingw` seems not needed for `sdist`?

Comment (guolinke) [inline python-package/setup.py]: 1. when building ```build_sdist```, we didn't need to compile.
2. yeah, ```--mingw```  is not needed in ```sdist``` .

**Comment by guolinke** [comment]
@wxchan maybe the ```bdist_wheel``` needs override as well ?

**Comment by wxchan** [comment]
@guolinke I found the reason here https://github.com/pypa/setuptools/blob/master/setuptools/command/install.py#L64. But I am not sure if it will cause other error.

**Comment by wxchan** [comment]
@guolinke Almost finished. There are two tests not passed yet:
1. how to set `OpenCL_INCLUDE_DIR` without `-DOpenCL_INCLUDE_DIR` for cmake? I use export OpenCL_INCLUDE_DIR, not working.
2. The wheel platform is not supported on travis, which I think it's macosx-10.7-x86_64.

**Comment by guolinke** [comment]
@wxchan 
Thanks, I think you can update the readme in python-package folder as well.

I think the  ```OpenCL_INCLUDE_DIR ``` may can be added in ```PATH```.

Not sure about ```BOOST_ROOT```.

**Comment by wxchan** [comment]
still cannot find `BOOST_ROOT`
```
 CMake Error at /usr/share/cmake-3.2/Modules/FindBoost.cmake:1201 (message):
      Unable to find the requested Boost libraries.
      Unable to find the Boost header files.  Please set BOOST_ROOT to the root
      directory containing Boost or BOOST_INCLUDEDIR to the directory containing
      Boost's headers.
```
@huanzhang12 do you know how to set `BOOST_ROOT` for cmake without `-DBOOST_ROOT`?

**Comment by wxchan** [comment]
Never mind, `export` seems working.

**Comment by guolinke** [comment]
@wxchan Thanks very much !
BTW, can you add following for the gpu installation from pip:
```
Boost and OpenCL are needed: details for installation can be found in gpu-support.
```

**Comment by wxchan** [comment]
@guolinke updated. btw, I didn't test `pip install lightgbm --install-option=--mingw`.

File: python-package/README.rst [hunk 2]
```diff
@@ -29,9 +29,17 @@ Install from source package:
...
 
-Note: Installation from source package require installing `cmake <https://cmake.org/>`_ first, and Visual Studio (or `MS Build <https://www.visualstudio.com/downloads/#build-tools-for-visual-studio-2017>`_) is needed in Windows. 
+Note: Installation from source package require installing `cmake <https://cmake.org/>`_ first, and Visual Studio (or `MS Build <https://www.visualstudio.com/downloads/#build-tools-for-visual-studio-2017>`_) is needed in Windows.
 
-For the MinGW build in Windows or GPU support, please install the latest version from GitHub.
+Install GPU version
+
+``pip install lightgbm --install-option=--gpu``
+
+Note: Boost and OpenCL are needed: details for installation can be found in `gpu-support <https://github.com/Microsoft/LightGBM/wiki/Installation-Guide#with-gpu-support>`_. Need to export `OpenCL_INCLUDE_DIR` and `BOOST_ROOT` before installation.
```
Comment (guolinke) [inline python-package/README.rst]: export OpenCL_INCLUDE_DIR ? or add it to the PATH ?

**Comment by guolinke** [comment]
@wxchan 
I add a small change: 
when building sdist, I think we need to copy the gpu files ( compute folder ) .

**Comment by wxchan** [comment]
@guolinke you also need to modify `.travis/test.sh`, `python setup.py sdist --gpu` will fail with unknown arguments. You can delete `initialize_options` function of `CustomSdist`.

**Comment by wxchan** [comment]
@guolinke what is the `_IS_SOURCE_PACKAGE.txt` for? It's created and deleted without copy_files in between. Is there any logic wrong?

**Comment by guolinke** [comment]
@wxchan 
This makes sure the built source package have ```_IS_SOURCE_PACKAGE.txt``` . 
So when installing from source package, it will not copy files. Otherwise it will always the newest cpp code then compile.

**Comment by wxchan** [comment]
@guolinke ok, get it.

**Comment by guolinke** [comment]
@wxchan 
pypi just move to https://pypi.org/project/lightgbm/ 
I am not sure how to update the document in that site.

**Pull request merged**